File: Strohm/ios/Strohm/Sources/Strohm/ViewModels/ViewModelBase.swift
import Foundation

open class ViewModelBase<DataType>: ObservableObject {
    var subscriptionId: UUID?
    let propName: PropName
    let propPath: PropPath
    @Published public var data: DataType

    public init(forPreviewWith data: DataType) {
        self.data = data
        self.propName = "use for preview only"
        self.propPath = []
    }

    public init(initialData: DataType, propName: PropName, propPath: PropPath) {
        self.data = initialData
        self.propName = propName
        self.propPath = propPath
        Strohm.default.subscribe(
            propsSpec: [propName: propPath],
            handler: receiveProps) { subscriptionId in
            self.subscriptionId = subscriptionId
        }
    }

    deinit {
        if let subscriptionId = self.subscriptionId {
            Strohm.default.unsubscribe(subscriptionId: subscriptionId)
        }
    }

    func receiveProps(props: Props) {
        if let data = propsToData(props: props) {
            DispatchQueue.main.async {
                self.data = data
                print("!!!!!!! did set data")
            }
        }
    }

    func propsToData(props: Props) -> DataType? {
        fatalError("abstract method")
    }
}

protocol PropsHandler {
    associatedtype Data

    func propsToData(props: Props) -> Data?
}

File: Strohm/android/strohm/src/main/java/dev/strohmnative/strohm/viewmodels/ViewModelBase.kt
package dev.strohmnative.strohm.viewmodels

import android.os.Handler
import android.os.Looper
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import dev.strohmnative.strohm.PropName
import dev.strohmnative.strohm.PropPath
import dev.strohmnative.strohm.Props
import dev.strohmnative.strohm.Strohm
import kotlinx.collections.immutable.persistentMapOf
import java.util.*

abstract class ViewModelBase<DataType>(
    initialData: DataType,
    val propName: PropName,
    val propPath: PropPath
) {
    private var subscriptionId: UUID? = null

    var data: MutableLiveData<DataType> =  MutableLiveStrohmData(::onActive, ::onInactive)
        private set

    private val handler = Handler(Looper.getMainLooper())

    init {
        data.value = initialData
    }

    private fun receiveProps(props: Props): Unit {
        propsToData(props)?.let { data ->
            handler.post(Runnable {
                this.data.value  = data
            })
        }
    }

    private fun onActive() {
        Strohm.getInstance().subscribe(persistentMapOf(propName to propPath), ::receiveProps) {
                subscriptionId -> this.subscriptionId = subscriptionId
        }
    }

    private fun onInactive() {
        this.subscriptionId = this.subscriptionId?.let {
            Strohm.getInstance().unsubscribe(it)
            null
        }
    }

    abstract fun propsToData(props: Props): DataType?
}

